###### 1) Загрузить и предварительно обработать данные:

Загрузить данные и проверить их корректность. Проверить наличие пропущенных значений и определить типы данных.

###### 2) Создать новый признак:

Ввести новый признак "марка автомобиля" (company). Идентифицировать производителей автомобилей, представленных в датасете. Внести необходимые коррективы в названия и проверить изменения.

###### 3) Выбрать предикторы и проанализировать корреляцию:

Оставить только ключевые предикторы. Оценить корреляцию между ценой (price) и другими переменными.

###### 4) Преобразовать категориальные переменные:

Использовать pd.get_dummies() для преобразования категориальных переменных.

###### 5) Построить модели линейной регрессии:

Создать модель с одним предиктором - мощность двигателя (horsepower). Оценить процент изменчивости, объясненной данной моделью (R^2).

###### 6) Сравнить модели:

Построить две модели: одну с использованием всех предикторов и вторую без учета марок машин. Сравнить изменения в (R^2), коэффициентах и их значимости. Принять решение о предпочтительной модели.

###### 7) Преобразовать категориальные переменные в фиктивные переменные (dummy variables).Использовать все доступные предикторы для построения модели.

Построение модели со всеми предикторами, кроме марок машин:

Исключить предикторы, относящиеся к маркам машин, и построить модель.

Оценка показателей всех трех моделей:

Вывести сводку результатов для каждой модели, включая R-squared, коэффициенты и значимость предикторов

1) Загрузить и предварительно обработать данные:

Загрузить данные и проверить их корректность. Проверить наличие пропущенных значений и определить типы данных.

In [9]:
import pandas as pd

df = pd.read_csv('cars.csv')
print(df.dtypes)
print(df.describe())
print(df.isna().sum())

car_ID                int64
symboling             int64
CarName              object
fueltype             object
aspiration           object
doornumber           object
carbody              object
drivewheel           object
enginelocation       object
wheelbase           float64
carlength           float64
carwidth            float64
carheight           float64
curbweight            int64
enginetype           object
cylindernumber       object
enginesize            int64
fuelsystem           object
boreratio           float64
stroke              float64
compressionratio    float64
horsepower            int64
peakrpm               int64
citympg               int64
highwaympg            int64
price               float64
dtype: object
           car_ID   symboling   wheelbase   carlength    carwidth   carheight  \
count  205.000000  205.000000  205.000000  205.000000  205.000000  205.000000   
mean   103.000000    0.834146   98.756585  174.049268   65.907805   53.724878   
std     59.3225

2) Использовать столбец CarName для формирования нового признака "марка автомобиля" (company).

Разбить значения ячеек по пробелу и записать в колонку первый элемент (например, 'audi 100 ls' → 'audi').

Подсчитать общее количество уникальных марок автомобилей в датасете.

Удалить столбец CarName с полным названием машины, так как использование полного наименования нецелесообразно.

Также удалить столбец car_ID, поскольку он не требуется для дальнейшего анализа.

In [46]:
import pandas as pd

df = pd.read_csv('cars.csv')

df['company'] = df.CarName.str.split(' ').apply(lambda x: x[0])
df = df.drop(['car_ID', 'CarName'], axis=1)

df.company.nunique()

28

В результате выполнения данного шага мы успешно создали новый признак "марка автомобиля" (company), используя полное название машины из столбца CarName. Общее количество уникальных марок машин в датасете равно 28. Затем мы удалили столбец car_ID и столбец CarName, так как они не будут использоваться в дальнейшем анализе.

3) Исправление ошибок в уникальных значениях

Исправить ошибки в названиях производителей автомобилей

Подсчитать количество уникальных производителей после коррекции.

Проанализировать, сколько уникальных производителей осталось в итоге.

In [51]:
import pandas as pd

df = pd.read_csv('cars.csv')

df['company'] = df.CarName.str.split(' ').apply(lambda x: x[0])
df = df.drop(['car_ID', 'CarName'], axis=1)

df['company'] = df.company.str.lower()
df['company'] = df.company.replace({'maxda': 'mazda', 'porcshce': 'porsche', 'toyouta': 'toyota', 'vokswagen': 'volkswagen', 'vw': 'volkswagen'})
df.company.nunique()

22

В результате коррекции обнаруженных ошибок в названиях производителей автомобилей мы успешно уменьшили количество уникальных производителей с 28 до 22.

4) Оставить только часть предикторов для предсказания цены:

'company', 'fueltype', 'aspiration','carbody', 'drivewheel', 'wheelbase', 'carlength','carwidth', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'boreratio','horsepower', 'price'.

Создать новый датафрейм с выбранными предикторами.

Посчитать корреляцию между ценой ('price') и остальными переменными в новом датафрейме.

In [76]:
import pandas as pd

df = pd.read_csv('cars.csv')

df['company'] = df.CarName.str.split(' ').apply(lambda x: x[0])
df = df.drop(['car_ID', 'CarName'], axis=1)

df['company'] = df.company.str.lower()
df['company'] = df.company.replace({'maxda': 'mazda', 'porcshce': 'porsche', 'toyouta': 'toyota', 'vokswagen': 'volkswagen', 'vw': 'volkswagen'})

df = df[['price', 'company', 'fueltype', 'aspiration','carbody', 'drivewheel', 'wheelbase', 'carlength','carwidth', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'boreratio','horsepower']]

corr_df = df.corr()
corr_df.sort_values(by=['price'], ascending=False).round(2)

price  wheelbase  carlength  carwidth  curbweight  enginesize  \
price        1.00       0.58       0.68      0.76        0.84        0.87   
enginesize   0.87       0.57       0.68      0.74        0.85        1.00   
curbweight   0.84       0.78       0.88      0.87        1.00        0.85   
horsepower   0.81       0.35       0.55      0.64        0.75        0.81   
carwidth     0.76       0.80       0.84      1.00        0.87        0.74   
carlength    0.68       0.87       1.00      0.84        0.88        0.68   
wheelbase    0.58       1.00       0.87      0.80        0.78        0.57   
boreratio    0.55       0.49       0.61      0.56        0.65        0.58   

            boreratio  horsepower  
price            0.55        0.81  
enginesize       0.58        0.81  
curbweight       0.65        0.75  
horsepower       0.57        1.00  
carwidth         0.56        0.64  
carlength        0.61        0.55  
wheelbase        0.49        0.35  
boreratio        1.00        0.57

В целом, корреляционные результаты подтверждают, что технические характеристики автомобилей, такие как мощность двигателя, размер двигателя, вес, ширина, длина, борирование и колесная база, существенно влияют на цену. Это позволяет сделать предположение о том, что более производительные и комфортабельные автомобили обычно имеют более высокую стоимость, что и очевидно.

5) Применение One-Hot Encoding:

Использовать функцию pd.get_dummies() для создания фиктивных переменных на основе изначальных категорий.

Применить One-Hot Encoding только к категориальным переменным (тип object).

Сравнить результаты с изначальными данными.

Применение One-Hot Encoding ко всему датафрейму:

Применить One-Hot Encoding ко всему датафрейму с тем же набором колонок, что и на предыдущем шаге.

In [75]:
import pandas as pd

df = pd.read_csv('cars.csv')

df['company'] = df.CarName.str.split(' ').apply(lambda x: x[0])
df = df.drop(['car_ID', 'CarName'], axis=1)

df['company'] = df.company.str.lower()
df['company'] = df.company.replace({'maxda': 'mazda', 'porcshce': 'porsche', 'toyouta': 'toyota', 'vokswagen': 'volkswagen', 'vw': 'volkswagen'})

df = df[['price', 'company', 'fueltype', 'aspiration','carbody', 'drivewheel', 'wheelbase', 'carlength','carwidth', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'boreratio','horsepower']]

df_dummy = pd.get_dummies(data=df[['company', 'fueltype', 'aspiration','carbody', 'drivewheel', 'wheelbase', 'carlength','carwidth', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'boreratio','horsepower', 'price']], drop_first = True)
df_dummy  

wheelbase  carlength  carwidth  curbweight  enginesize  boreratio  \
0         88.6      168.8      64.1        2548         130       3.47   
1         88.6      168.8      64.1        2548         130       3.47   
2         94.5      171.2      65.5        2823         152       2.68   
3         99.8      176.6      66.2        2337         109       3.19   
4         99.4      176.6      66.4        2824         136       3.19   
..         ...        ...       ...         ...         ...        ...   
200      109.1      188.8      68.9        2952         141       3.78   
201      109.1      188.8      68.8        3049         141       3.78   
202      109.1      188.8      68.9        3012         173       3.58   
203      109.1      188.8      68.9        3217         145       3.01   
204      109.1      188.8      68.9        3062         141       3.78   

     horsepower    price  company_audi  company_bmw  ...  enginetype_ohc  \
0           111  13495.0             0            0  ...               0   
1           111  16500.0             0            0  ...               0   
2           154  16500.0             0            0  ...               0   
3           102  13950.0             1            0  ...               1   
4           115  17450.0             1            0  ...               1   
..          ...      ...           ...          ...  ...             ...   
200         114  16845.0             0            0  ...               1   
201         160  19045.0             0            0  ...               1   
202         134  21485.0             0            0  ...               0   
203         106  22470.0             0            0  ...               1   
204         114  22625.0             0            0  ...               1   

     enginetype_ohcf  enginetype_ohcv  enginetype_rotor  cylindernumber_five  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                1                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    1   
..               ...              ...               ...                  ...   
200                0                0                 0                    0   
201                0                0                 0                    0   
202                0                1                 0                    0   
203                0                0                 0                    0   
204                0                0                 0                    0   

     cylindernumber_four  cylindernumber_six  cylindernumber_three  \
0                      1                   0                     0   
1                      1                   0                     0   
2                      0                   1                     0   
3                      1                   0                     0   
4                      0                   0                     0   
..                   ...                 ...                   ...   
200                    1                   0                     0   
201                    1                   0                     0   
202                    0                   1                     0   
203                    0                   1                     0   
204                    1                   0                     0   

     cylindernumber_twelve  cylindernumber_two  
0                        0                   0  
1                        0                   0  
2                        0                   0  
3                        0                   0  
4                        0                   0  
..                     ...                 ...  
200                      0                   0  
201      

Результаты One-Hot Encoding применены к категориальным переменным. Структура данных изменилась, и теперь включает фиктивные переменные.

6) Использовать библиотеку statsmodels для построения линейной регрессии с одним предиктором – horsepower.

Рассчитать процент изменчивости (R^2) для полученной модели и округлить до целого числа.

In [89]:
import pandas as pd
import statsmodels.formula.api as smf 

df = pd.read_csv('cars.csv')

df['company'] = df.CarName.str.split(' ').apply(lambda x: x[0])
df = df.drop(['car_ID', 'CarName'], axis=1)

df['company'] = df.company.str.lower()
df['company'] = df.company.replace({'maxda': 'mazda', 'porcshce': 'porsche', 'toyouta': 'toyota', 'vokswagen': 'volkswagen', 'vw': 'volkswagen'})

df = df[['price', 'company', 'fueltype', 'aspiration','carbody', 'drivewheel', 'wheelbase', 'carlength','carwidth', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'boreratio','horsepower']]

result = smf.ols('price ~ horsepower', data=df).fit()

print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.653
Model:                            OLS   Adj. R-squared:                  0.651
Method:                 Least Squares   F-statistic:                     382.2
Date:                Tue, 19 Sep 2023   Prob (F-statistic):           1.48e-48
Time:                        19:32:14   Log-Likelihood:                -2024.0
No. Observations:                 205   AIC:                             4052.
Df Residuals:                     203   BIC:                             4059.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -3721.7615    929.849     -4.003      0.0

Модель с одним предиктором horsepower объясняет примерно 65% изменчивости в цене автомобилей. Это говорит о том, что мощность двигателя значительно влияет на стоимость автомобиля.

7) Использовать библиотеку statsmodels для построения линейной регрессии с одним предиктором – horsepower.

Построение модели со всеми предикторами:

Преобразовать категориальные переменные в фиктивные переменные (dummy variables).Использовать все доступные предикторы для построения модели.

Построение модели со всеми предикторами, кроме марок машин:

Исключить предикторы, относящиеся к маркам машин, и построить модель.

Оценка показателей всех трех моделей:

Вывести сводку результатов для каждой модели, включая R-squared, коэффициенты и значимость предикторов.

In [129]:
import pandas as pd
import statsmodels.api as sm

df = pd.read_csv('cars.csv')

df['company'] = df.CarName.str.split(' ').apply(lambda x: x[0])
df = df.drop(['car_ID', 'CarName'], axis=1)

df['company'] = df.company.str.lower()
df['company'] = df.company.replace({'maxda': 'mazda', 'porcshce': 'porsche', 'toyouta': 'toyota', 'vokswagen': 'volkswagen', 'vw': 'volkswagen'})
df = df[['price', 'company', 'fueltype', 'aspiration','carbody', 'drivewheel', 'wheelbase', 'carlength','carwidth', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'boreratio','horsepower']]
df_dummy = pd.get_dummies(data=df[['company', 'fueltype', 'aspiration','carbody', 'drivewheel', 'wheelbase', 'carlength','carwidth', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'boreratio','horsepower', 'price']], drop_first = True)


X = df_dummy.drop(columns=['price'])
Y = df_dummy['price']  
X = sm.add_constant(X)  
model = sm.OLS(Y, X)  
results = model.fit()  
print(results.summary())

X = df_dummy.drop(columns=['price', 'company_audi', 'company_bmw', 'company_buick', 'company_chevrolet', 'company_dodge', 'company_honda', 'company_isuzu', 'company_jaguar', 'company_mazda', 'company_mercury', 'company_mitsubishi', 'company_nissan', 'company_peugeot', 'company_plymouth', 'company_porsche', 'company_renault', 'company_saab', 'company_subaru', 'company_toyota', 'company_volkswagen', 'company_volvo'])
Y = df_dummy['price']  
X = sm.add_constant(X)  
model = sm.OLS(Y, X)  
results = model.fit()  
print(results.summary())



                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.959
Model:                            OLS   Adj. R-squared:                  0.948
Method:                 Least Squares   F-statistic:                     81.09
Date:                Tue, 19 Sep 2023   Prob (F-statistic):           4.86e-89
Time:                        20:08:04   Log-Likelihood:                -1804.2
No. Observations:                 205   AIC:                             3702.
Df Residuals:                     158   BIC:                             3858.
Df Model:                          46                                         
Covariance Type:            nonrobust                                         
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
const                 -3.472e+

Модель с одним предиктором (horsepower):

R^2=0.653: Мощность двигателя (horsepower) объясняет примерно 65% изменчивости цены автомобилей.

Модель со всеми предикторами:

R^2=0.959: Модель, включающая все предикторы, в том числе и марки машин, хорошо объясняет более 95% изменчивости цены.

Модель со всеми предикторами, кроме марок машин:

R^2=0.914: Этот вариант модели, исключая марки машин, также обладает высоким R^2, равным примерно 91%.


Таким образом, можно сделать вывод, что технические характеристики автомобилей, особенно мощность двигателя, существенно влияют на цену.